### ISSUE TITLE
Panoramax image upload failing on Chrome Android

### ISSUE THREAD

**Issue by davidpnewton** [pietervdvn/MapComplete#2202]
Using 0.47.5 on Android with Chrome and when trying to upload images to Panoramax I'm getting a failed upload message. Same message as was showing up when Imgur image uploads started failing.

0.47.5 on Windows with Chrome allows upload of images to Panoramax and so it appears to be an Android-specific Chrome issue.

**Comment by thibaultmol** [comment]
Yeah, we are aware of the issue but still trying to figure it out.

From my understanding and debugging from it. It seems like on some phones, if a browser opens an 'image picker', you select the image, try upload. it'll fail. (and then the error behind the scenes will say it's because there is no location in the image). Because some phones strip location from images it seems.

Could you try this workaround:
when you click the button to upload a file, click the top right menu and choose browse. and select the image that way

![image](https://github.com/user-attachments/assets/805a8e63-a34e-40de-8199-937c7d199caa)

![image](https://github.com/user-attachments/assets/717b0e3e-422a-4a0a-b9a3-4a87b861285f)

![image](https://github.com/user-attachments/assets/17f0d699-958f-4d00-a4b5-9c70c26eba5d)

**Comment by pietervdvn** [comment]
@davidpnewton Would you mind sharing:

- Your android version
- The chrome version you are using
- Your OSM-username (I have logging and have some logs of failed uploads)

If you don't want to share this publicly, please send this to me via [the osm messaging system](https://www.openstreetmap.org/user/Pieter%20Vander%20Vennet) or some other means.

**Comment by RayBB** [comment]
I recently had a similar issue on my app. On Android I think all browsers strip location if the html of the page is set to upload images. If you set it to upload any file type (or something like that) then it stops stripping location. Tested this on FF and Chrome Android.

**Comment by thibaultmol** [comment]
also what I observed. using a 'file picker' instead of an 'image picker' solves this problem from my testing.
Problem is that the file picker doesn't have an easy 'take picture' button like the image picker does.

@pietervdvn I think maybe settling for the file picker is best. And when you have you have the android package, you can then use the native android image picker which won't have this problem https://github.com/pietervdvn/MapComplete/issues/2112

**Comment by pietervdvn** [comment]
I tried this with Google Chrome on an older device and could not reproduce this.

@davidpnewton I did make some fixes on dev.mapcomplete.org though, which might fix this. Would you mind testing there too?

**Comment by davidpnewton** [comment]
Just did. I'm afraid it didn't resolve the issue with the image picker. The file browser method does function as a workaround.

So it does indeed appear to be something funky that the image picker is doing to the pictures v what the file browser does to them.

**Comment by davidpnewton** [comment]
It's Firefox as well that shows this bug on Android. Definitely looks like an image picker problem. Devtools on Firefox showed this:

![Screenshot_20241010_203505_Firefox](https://github.com/user-attachments/assets/3d81f7ca-64a6-42b3-9899-8c149b44dcb6)

**Comment by pietervdvn** [comment]
Allright, first a mea culpa. I never actually pushed the changes...

Second, I'm not sure the EXIF-data is the problem - the only case that I didn't cover (up till just now) is that if there is _no_ exif data at all and that the EXIF-library crashes on this. I put a try-catch around it to mitigate this, so it might help.

Furthermore, I've added in error reporting code, so it should now "phone home" with the stack trace, so I can peek into the logs without you needing to look peek into the console ;)

BTW: is that some plugin which gives a console in the mobile browser? Which one is it?

**Comment by davidpnewton** [comment]
"Allright, first a mea culpa. I never actually pushed the changes..."

Oopsie. I'll try again now that you have, really, pushed them!

As for the console on a mobile browser? That's a screenshot of Firefox, and there's a Firefox addon which gives a console for the mobile version.

**Comment by thibaultmol** [comment]
Oh why didn't this come up when I searched for it before xd really cool: https://addons.mozilla.org/en-US/android/addon/mobidevtools/

**Comment by thibaultmol** [comment]
v0.47.5 fully cleared data before hand:
![Screenshot_20241011-124518](https://github.com/user-attachments/assets/c29ad180-c9bd-49b8-86e1-b8b44cfac75f)
![Screenshot_20241011-124510](https://github.com/user-attachments/assets/b3726202-dcc7-42a0-b159-025dca6db8d0)

**Comment by Binnette** [comment]
Same problem on my Android phone with Chrome (or Firefox).
- Osm user : Binnette
- Upload datetime : Mon, 21 Oct 2024 07:59:39 GMT

POST on https://panoramax.mapcomplete.org/api/collections/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/items/
returns: 
````json
{
    "details": {
        "error": "Coordonn\u00e9es GPS absentes ou invalides dans les attributs EXIF de l'image"
    },
    "message": "Impossible de lire les m\u00e9tadonn\u00e9es de la photo",
    "status_code": 400
}
````
The error is "GPS coors missing or invalid in image EXIF".

So I tried to upload the same image but with Firefox on PC and it works nicely.

So I recorded a "HAR" from my phone and a second "HAR" with the web browser of my PC. Both contains the payload and the image to upload on panoramax. But the image binary are different for the same image (phone vs PC)! I can not compare more, I didn't find a way to properly extract the image from the HAR file...

Also I noticed on phone that:
- If I use "camera" (1) directly from MapComplete, the upload works
- If I use "Media file (2) selector" from MapComplete, the upload fail.

![image](https://github.com/user-attachments/assets/77b78242-ba83-435c-be44-1f51d80f4cc3)

**Comment by thibaultmol** [comment]
@Binnette do the instructions for using the 2nd option you show work that I showed in the 2nd comment on this issue? 
clicking the triple dot menu at the top right and choosing browse and selecting the file that way?

**Comment by Binnette** [comment]
> @Binnette do the instructions for using the 2nd option you show work that I showed in the 2nd comment on this issue? clicking the triple dot menu at the top right and choosing browse and selecting the file that way?

Ok it works! Sorry I didn't understand that this procedure was a work around. TY @thibaultmol

**Comment by davidpnewton** [comment]
https://stackoverflow.com/questions/70865878/picking-photo-from-web-in-android-no-longer-includes-exif-gps-location

This may add a little information about what's going on.

**Comment by thibaultmol** [comment]
yeah that's what I said to pieter as well. I'm not sure why they're doing that but yeah.

I suggested that when there is a mapcomplete android app (just a wrapper around the website) he'd be able to get around this problem.

and on the regular website just switch to using the regular file picker maybe

**Comment by pietervdvn** [comment]
> yeah that's what I said to pieter as well. I'm not sure why they're doing that but yeah.

Privacy

@Binnette I'd love to receive the HAR-files, especially of the failing upload! You can send it to info@mapcomplete.org

**Comment by thibaultmol** [comment]
> Privacy

Well yeah, but then give an option to the user to disable it stripping it. (I was thinking just now like: why is the camera working for Binnette and then I realized why: it's the samsung galaxy camera app which probably just doesn't do the gps coordinate stripping)

**Comment by pietervdvn** [comment]
I did an attempt that might improve matters here. @Binnette can you please try again?

**Comment by Binnette** [comment]
Hi @pietervdvn, without using the workaround I still have the issue on dev instance.

**Comment by Hedaja** [comment]
> I did an attempt that might improve matters here. @Binnette can you please try again on https://dev.mapcomplete.org ?

I unfortunately can confirm that the issue still persists on the latest dev version

Tried it on Android 11 with Firefox 133.0b9.

**Comment by RayBB** [comment]
![Screenshot_20241124-181755](https://github.com/user-attachments/assets/fc997c8c-a002-45a3-b5a5-ff7474e8fa73)
Firefox on a pixel 7 android 14

**Comment by pietervdvn** [comment]
@RayBB Seems like the test here works fine, but you cannot upload pictures linked to a feature? Would you mind testing that?

**Comment by Binnette** [comment]
Test page work great. I click select file, it open the "Open from" intent:
![Screenshot_20241125_015650_Files](https://github.com/user-attachments/assets/73f30374-bc09-4a3d-9bd2-5b5d2a16c3b4)

I can pick a picture and upload start (no confirm button to press or whatever)

![Screenshot_20241125_022644_Firefox](https://github.com/user-attachments/assets/d4ceced1-62e4-4b2b-856c-a5bf1e22dac8)

**Comment by pietervdvn** [comment]
But for both of you, uploading linked to a feature does _not_ work? Then something weird is going on

**Comment by Binnette** [comment]
Here is a way to reproduce the "bad/broken" behavior of Map Complete add a photo button.

Use the select file of test page. It open the "Open from" intent (note that I can choose apps like gallery or the "media picker" used by map complete button)
![Screenshot_20241125_015650_Files](https://github.com/user-attachments/assets/8c3efa83-7d91-49f0-8df7-b8ec66bdb378)

So I choose media picker, it show this view:
![Screenshot_20241125_023245_Media picker](https://github.com/user-attachments/assets/d6295c66-7fcf-41c5-8a0b-200ed2b91868)

I choose a picture and upload start but this time it fail:
![Screenshot_20241125_023406_Firefox](https://github.com/user-attachments/assets/c599f490-bd2e-40d6-b2f7-be80273e5f80)

So issue comes from the "media picker". Map Complete should use the default general file picker and the the bugguy image/media picker with file extension filters

**Comment by Binnette** [comment]
When I click on map complete add a picture button, this view is show:
![Screenshot_20241125_015530_IntentResolver](https://github.com/user-attachments/assets/b0ab0d67-738c-479e-8371-e280a3bbd18a)

Since I want to upload an picture from my phone, I choose "media picker". Here is the source of our problem -> the media picker

**Comment by thibaultmol** [comment]
(as i've been saying, exactly).
it seems like on certain android devices, media picker strips gps coordinates (maybe for privacy reasons?)

**Comment by pietervdvn** [comment]
Oooohhhh, you get a NaN-value.

**Comment by pietervdvn** [comment]
Allright, I pushed a fix which might finally resolve this issue. When [this build](https://github.com/pietervdvn/MapComplete/actions/runs/12007802525) is completed (and about ten more minutes later), do a hard refresh of mapcomplete and please test again to upload a picture using the map view. If it doesn't work, use the debug-view and take a screenshot of what is happening.

Thanks to the people testing here! I couldn't reproduce it locally which made it very hard to debug.

**Comment by Binnette** [comment]
I agree @thibaultmol. I have Samsung device with Android 14. Either it concerns certain brands of phone, or either it concerns some specific Android versions.

**Comment by pietervdvn** [comment]
Yeah, this is tied to specific version of Android in combination with certain media pickers - part of the reason this was so hard for me to debug.

**Comment by Binnette** [comment]
I just tried on MapComplete I still have to choose action between camera and media picker. Of course when I choose media picker it fails uploading the picture.

I tested also the test page. And I force myself to choose media picker. My picture is uploaded with a weird location is Belgium (my picture was took in France)

![Screenshot_20241125_120619_Firefox](https://github.com/user-attachments/assets/0055462d-b940-4757-81fd-294cd2e99348)

I think the issue is the media picker. We should not use it in MapComplete as it is bugguy. And instead use the "file" picker.

> The new photo picker has the potential to strip location info from EXIF

More info here: https://github.com/commons-app/apps-android-commons/issues/5242#issue-1762403088

**Comment by pietervdvn** [comment]
I've indeed also changed the accepted types from "image/*" to ".jpg, .jpeg" only. This too might have caused a different file picker to open. 

I'm gonna revert this, cause I want people to be able to use the camera and move this behaviour to a second button (which is pretty much the implementation that @thibaultmol was nagging me for)

**Comment by Binnette** [comment]
Yes maybe 2 buttons can be a good "work around/solution" :
- "📸 Take a photo" (instead of "Add picture"). It is more explicit that the action is to "take a picture" now.
- "📁 Add image from gallery" (instead of "Use the file selector dialog").

**Comment by thibaultmol** [comment]
I also suggested: 
Have mapcomplete detect on the first image upload if the current device is indeed stripping coordinates or not.
If it's stripping them, change something for the future uploads on that device (using a cookie or something to store this info).
Make the file picker button more prominent. or somehting.

**Comment by pietervdvn** [comment]
Ok, so, the current interface (which is building now) will have two buttons:

One is the well-known, big camera button. This one will prompt to take a picture right away or will use the _media_ picker (which might strip the coordinates). If a user picks this, mapcomplete will first try to use the EXIF-data, then the GPS location, then the feature centerpoint. The assumption here is that the contributor adds it in place.

The second, smaller button opens the _file_ picker directly. It still will only show jpg-pictures though. This one will use the EXIF-data and, if not available, the features centerpoint. _This method does not use the current GPS location as it assumes that the pictures were taken a long time before at a different location_.

Please test this and give some feedback.

**Comment by thibaultmol** [comment]
I like this a lot! (with a thing built in that remembers the choice on that device I suppose. and/or for a set period

**Comment by pietervdvn** [comment]
Every additional button and step is problematic. Try giving the app to your a non-technical person. They will be confused about the two buttons and ask what the difference is. I'm already not happy with the extra button.

An additional question aobut "what coordinates that should be used" will be confusing too. This will provoke reactions as "why does this picture need a location", "doesn't the device already know the location", ... _You guys are the power users of this tool. Most people using MapComplete have no idea what they are doing_

Thirdly, every feature and UI is also something that needs to be maintained.

At last, I personally don't care too much about having the precise geolocation. The data in panoramax.mapcomplete.org is already heavily polluted by centerpoints anyway. I mostly care about the pictures being safe and available; using Panoramax gives a lot of extra benefits, but the core task is to keep the pictures safe.

As such, having an extra step as Binette proposes is something I'd only do if really, really necessary.

However, a compromise might be possible. For example, if EXIF-data is not available and if the GPS-location is 50m away from the feature (or the features edge); or if the timestamp is a certain amount older then e.g. 15 minutes, I might default to using the centerpoint instead of the geolocation which is available to the webapp.

**Comment by davidpnewton** [comment]
Further problem with the Panoramax instance: it appears that the locations for recent photos are utter garbage.

ID a67bf392-baf2-4af9-b0e8-a76c7453171e should be in the south of Leamington Spa in Warwickshire. It actually shows up in Suffolk, hundreds of miles east of where it supposed to be!

**Comment by pietervdvn** [comment]
@davidpnewton This is because of #2304 . I'm marking this comment as "irrelevant" as this is about a different (but related) issue.
